File: leetcode/src/partitionpalindromeII.cpp
/*
 * given a string s, partition s such that every substring of partition is a palindrome
 * return the minimum cuts needed for a palindrome partitioning of s
 *
 * test data:
 * cbbcc
 * aaaaaaaaaab
 * */

#include <string>
#include <cstring>
#include <algorithm>

using namespace std;

class Solution{

public:
    /*
     *
     * */
    int minCut(const string& s){
        const int n = s.size();
        if(n < 2)        return 0;

        int dp[n][n+1];    // dp[i][j] == 1 means s.substr(i,j-i) is palindrome, both of i and j are char index
        for(int i = 0; i < n; ++i){
            memset(dp[i], 0, sizeof(dp[i]));
            dp[i][i+1] = 1;
        }

        for(int len = 2; len <= n; ++len){
            for(int start = 0; start + len <= n; ++start){
                if(len < 4 || dp[start + 1][start + len - 1] == 1){
                    if(s[start] == s[start + len - 1]){
                        dp[start][start + len] = 1;
                    }
                }
            }
        }

        int mc[n+1];  // mc[i] means min cuts of substring [0,i)
        memset(mc, 0, sizeof(mc));

        for(int e = 2; e <= n; ++e){
            if(dp[0][e] == 1){
                mc[e] = 0;
            }else{
                mc[e] = mc[e-1] + 1;   // default value
                for(int j = e-1; j > 0; --j){ // iterate j backward to make use of previously fixed mc[j]
                    if(dp[j][e] == 1){
                        mc[e] = min(mc[e], mc[j] + 1);
                    }
                }
            }
        }

        return mc[n];
    }
};

/* unit test is in ../cpp_unittest/partitionpalindromeII_unittest */

File: leetcode/java/partitionpalindromeII.java
/*
 * given a string s, partition s such that every substring of the partition is a palindrome. 
 * return the minimum cuts needed for a palindrome partition of s.
 * 
 * test data:
 * a
 * abb
 * abad
 * aaaaaaaaa
 * aaaaaaaaaaaaab
 * baaaaaaaaaaaa
 * */
import java.io.*;
import java.lang.Math;
import java.util.Arrays;

public class partitionpalindromeII{
    /*
     * accepted solution on oj.leecode. totally, time O(n^2) and space O(n^2)
     * */
    public int minCut(String s){
        if(s == null || s.length() < 2)    return 0;
        char[] arr = s.toCharArray();
        int n = arr.length;

        boolean[][] dp = new boolean[n][n+1];
        for(int i = 0; i < n; ++i){
            for(int j = i + 2; j <= n; ++j){
                dp[i][j] = false;
            }
            dp[i][i] = true;
            dp[i][i+1] = true;
        }

        for(int l = 2; l <= n; ++l){ // check all substr palindrome information 
            for(int e = n; e - l >= 0; --e){
                if(l < 4 || dp[e - l + 1][e - 1]){
                    if(arr[e - l] == arr[e - 1]){
                        dp[e - l][e] = true;
                    }
                }
            }
        }

        int[] mcl = new int[n + 1]; // mcl[l] means min cut of substring [0, e)
        for(int e = 2; e <= n; ++e){ // index end
            if(dp[0][e]){
                mcl[e] = 0;
            }else{
                int in = e - 2; // init value is maximum cut of l-1 except 1 cut to split [0,j) and [j, e)
                for(int j = e - 1; j > 0; --j){ // j must start from l-1! as dp[l-1][l]==1, only at time, in can be minimized to mcl[l-1]
                    if(dp[j][e])    in = Math.min(in, mcl[j]);
                }
                mcl[e] = in + 1;
            }
        }
        return mcl[n];
    }

    /* test case is in  ../partitionpalindromII_junit/ */
}
